HACKING.md

```diff
@@ -520,6 +520,228 @@ Common instructions for porting:
+       boolean check on the number, like `numItems > x`.  In this case you
+       should just pass the result of `numItems > x` as a prop.
+
+    3. Make liberal use of
+       [React.memo](https://reactjs.org/docs/react-api.html#reactmemo),
+       especially for components that have complicated render functions.
+       Assuming you also follow the previous point (passing only the
+       essential information used by the component), then `React.memo` will
+       ensure that rendering can be skipped when the props don't change.
+       The cost of React comparing the props for a small sub-tree will be
+       much less than the cost of rendering the entire tree again.
+
+    It's highly recommended to install the
+    [React Developer Tools](https://beta.reactjs.org/learn/react-developer-tools)
+    browser extension to indentity areas where components are updating
+    unnecessarily.  With the tools pane open, click on the settings cog
+    and make sure the following checkboxes are enabled:
+
+      * "Highlight updates when components render," under "General."
+
+      * "Record why each component rendered while profiling," under
+         "Profiler."
+
+    The first setting will highlight component updates on the page, which
+    should make it obvious where unnecessary updates are occurring. (For
+    example, editing a relationship shouldn't update every other relationship
+    on the page.) The second setting will tell you in the profiler which
+    props changed to trigger those updates.
+
+    Once you indentify which props are triggering unnecessary updates,
+    it's generally just a matter of making them more specific as per above,
+    or caching them with the `useMemo` and `useCallback` hooks. Another
+    situation might be where you are passing a prop which is used in a
+    callback (say, an event handler), but doesn't actually affect the
+    rendered output at all. Consider passing that prop as a
+    [ref](https://reactjs.org/docs/hooks-reference.html#useref) instead,
+    since the ref's identity will remain stable.
+
+  * The second issue with top-down data flow is related to organization: if
+    there's only one reducer function, then how do we avoid it becoming a
+    mess, managing tons of actions from many different files? And how do we
+    make it clear what component an action is dispatched from?
+
+    Even though there's only one "real" reducer, we don't have to make it
+    handle every single action in the app on its own. We can still define
+    separate reducers for each child component, and simply call them from the
+    parent reducer.
+
+    A general outline of the types and functions you'd typically define in
+    a component file are described below. These are specific to the file's
+    component, and handle all the actions/state used by that component. The
+    parent doesn't have to worry about the child's actions unless it needs
+    to.
```

Comment:
It doesn't have to unless it has to? :) What do you have in mind?

Reply:
Basically what is explained further down for `update-child`:

```js
          /*
           * You could even have another switch statement here on
           * childAction.type, in case you need to handle particular actions
           * in the parent reducer (because they affect state in the parent
           * that the child doesn't know about).
           */
```
But I'll just remove this sentence because it's confusing and doesn't add anything.